lib/pg_eventstore/queries.rb

```diff
@@ -0,0 +1,140 @@
+    def initialize(connection, serializer, deserializer)
+      @connection = connection
+      @serializer = serializer
+      @deserializer = deserializer
+    end
+
+    # @return [void]
+    def transaction
+      connection.with do |conn|
+        # We are inside a transaction already - no need to start another one
+        if [PG::PQTRANS_ACTIVE, PG::PQTRANS_INTRANS].include?(conn.transaction_status)
+          next yield
+        end
+
+        conn.transaction do
+          conn.exec("SET TRANSACTION ISOLATION LEVEL SERIALIZABLE")
+          yield
+        end
+      end
+    rescue PG::TRSerializationFailure, PG::TRDeadlockDetected => e
+      retry if e.connection.transaction_status == PG::PQTRANS_IDLE
+      raise
+    end
+
+    # Finds a stream in the database by the given Stream object
+    # @param stream [PgEventstore::Stream]
+    # @return [PgEventstore::Stream, nil] persisted stream
+    def find_stream(stream)
+      find_sql = <<~SQL
+        SELECT * FROM streams WHERE streams.context = $1 AND streams.stream_name = $2 AND streams.stream_id = $3
+          LIMIT 1
+      SQL
+      pgresult = connection.with do |conn|
+        conn.exec_params(find_sql, stream.to_a)
+      end
+      PgEventstore::Stream.new(**pgresult.to_a.first.transform_keys(&:to_sym)) if pgresult.ntuples == 1
+    end
+
+    # @param stream [PgEventstore::Stream]
+    # @return [PgEventstore::RawStream] persisted stream
+    def create_stream(stream)
+      create_sql = <<~SQL
+        INSERT INTO streams (context, stream_name, stream_id) VALUES ($1, $2, $3) RETURNING *
+      SQL
+      pgresult = connection.with do |conn|
+        conn.exec_params(create_sql, stream.to_a)
+      end
+      PgEventstore::Stream.new(**pgresult.to_a.first.transform_keys(&:to_sym))
+    end
+
+    # @return [PgEventstore::Stream] persisted stream
+    def find_or_create_stream(stream)
+      find_stream(stream) || create_stream(stream)
+    end
+
+    # @see PgEventstore::Client#read for more info
+    # @param stream [PgEventstore::Stream]
+    # @param options [Hash]
+    # @return [Array<PgEventstore::Event>]
+    def stream_events(stream, options)
+      exec_params = events_filtering_builder(stream, options).to_exec_params
+      pgresult = connection.with do |conn|
+        conn.exec_params(*exec_params)
+      end
+      deserializer.deserialize_many(pgresult)
+    end
+
+    # @param stream [PgEventstore::Stream] persisted stream
+    # @param event [PgEventstore::Event]
+    # @return [PgEventstore::Event]
+    def insert(stream, event)
+      serializer.serialize(event)
+
+      attributes = event.options_hash.slice(:id, :type, :data, :metadata, :stream_revision, :link_id).compact
+      attributes[:stream_id] = stream.id
+
+      sql = <<~SQL
+        INSERT INTO events (#{attributes.keys.join(', ')}) 
+          VALUES (#{(1..attributes.values.size).map { |n| "$#{n}" }.join(', ')}) 
+          RETURNING *
+      SQL
```

Comment:
This should be put in its own method / class.

Reply:
Yes. It could be another SQL builder that computes an `INSERT` query. I already implemented such builder for `SELECT` queries - `lib/pg_eventstore/sql_builder.rb`. Currently `INSERT` statements are not complex, so at this point I didn't want to spend a time on implementing a builder for it

Reply:
i dont mind if you leave it here, but you could move it to its own method at least, to improve readability.

Reply:
Moving it into a separated method won't make it more readable, because it will still be the same string with the same number of variables in there. As I wrote - I'll have to implement the `INSERT` query builder